File: bot.scala
import scala.annotation.tailrec
import scala.util.Try
import sys.env

def expect[T](get: String => T, key: String, hint: String): T =
  Try(get(key)).getOrElse(sys.error(s"Bad or missing $key in $hint"))

class Git(val worktree: os.Path):
  def clone(repo: String, branch: String, user: String): Unit =
    os.proc("git", "clone", "--depth", "1", "--branch", branch, repo, worktree.toString)
      .call(stdout = os.Inherit)
    os.proc("git", "config", "user.email", user).call(cwd = worktree, stdout = os.Inherit)
    os.proc("git", "config", "user.name", user).call(cwd = worktree, stdout = os.Inherit)
  def add(sub: String, files: Seq[(String, String)]): Unit =
    os.proc("git", "pull").call(cwd = worktree, stdout = os.Inherit)
    for (name, content) <- files do os.write.over(worktree / sub / name, content)
    os.proc("git", "add", ".").call(cwd = worktree, stdout = os.Inherit)
    os.proc("git", "commit", "-m", "Sync updates").call(cwd = worktree, stdout = os.Inherit)
    os.proc("git", "push").call(cwd = worktree, stdout = os.Inherit)

def getUpdates(token: String, offset: Long): Seq[ujson.Value] =
  val url = s"https://api.telegram.org/bot$token/getUpdates"
  val params = Map("timeout" -> "25", "offset" -> offset.toString)
  val resp = requests.get(url, params = params, readTimeout = 30_000, connectTimeout = 30_000)
  if resp.statusCode != 200 then sys.error(s"getUpdates failed with HTTP ${resp.statusCode}")
  val payload = ujson.read(resp.text())
  if !payload("ok").bool then sys.error("Telegram getUpdates returned failure")
  payload("result").arr.toSeq

@main def theMain(): Unit =
  val confObj = expect(k => ujson.read(env(k)), "TG2SITE_CONF_CONTENT", "env")
  val secretsObj = expect(k => ujson.read(os.read(os.Path(env(k)))), "TG2SITE_SECRETS_PATH", "env")
  val repoURL = expect(secretsObj(_).str, "repository_url", "secret")
  val publishBranch = expect(confObj(_).str, "publish_branch", "conf")
  val channelID = expect(confObj(_).num.toLong, "channel_id", "conf")
  val telegramToken = expect(secretsObj(_).str, "telegram_token", "secret")

  val git = new Git(os.temp.dir(prefix = "tg2site-") / "repo")
  System.err.println(s"cloning repo=$repoURL branch=$publishBranch into ${git.worktree}")
  git.clone(repoURL, publishBranch, "bot@tg2site")
  val messagesDir = git.worktree / ".tg2site-messages"
  os.makeDir.all(messagesDir)
  val messageKeys = Seq("channel_post", "edited_channel_post", "message", "edited_message")
  @tailrec def iteration(lastOffset: Long): Unit =
    val updates = getUpdates(telegramToken, lastOffset + 1)
    val messagesByChat = (for
      update <- updates
      messageKey <- messageKeys
      message <- update.obj.get(messageKey)
      chatId <- Try(message("chat")("id").num.toLong).toOption
      path = f"${update("update_id").num.toLong}%020d.$messageKey.json"
    yield chatId -> (path -> ujson.write(message))).groupMap(_._1)(_._2)
    System.err.println(s"got updates from: ${messagesByChat.keys.toSeq.sorted}")
    for relevant <- messagesByChat.get(channelID) do git.add(".tg2site-messages", relevant)
    iteration(if updates.isEmpty then lastOffset else updates.last("update_id").num.toLong)
  val NumP = """.*/(\d+).*\.json""".r
  val messages = os.list(messagesDir)
  iteration(messages.map(_.toString).collect{ case NumP(s) => s.toLong }.maxOption.getOrElse(-1))

/***
`System.err` usage is intentional
***/

File: bot.java
//import com.fasterxml.jackson.databind.ObjectMapper;
//import com.fasterxml.jackson.databind.node.ArrayNode;

//import tools.jackson.core.

import tools.jackson.databind.ObjectMapper;
import tools.jackson.databind.node.ArrayNode;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Duration;
import java.util.*;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;

// error util

interface RUCallable<T> {
    T call() throws Exception;
}

<T> T unchecked(RUCallable<T> body) {
    try {
        return body.call();
    } catch (Exception e) {
        throw (e instanceof RuntimeException re ? re : new RuntimeException(e));
    }
}

<R extends AutoCloseable, T> T using(Supplier<R> resource, Function<R, T> body) {
    try (R r = resource.get()) {
        return body.apply(r);
    } catch(Exception e){
        throw (e instanceof RuntimeException re ? re : new RuntimeException(e));
    }
}

<T> T expect(Function<String,T> get, String key, String hint){
    try {
        return get.apply(key);
    } catch (Exception e) {
        throw new RuntimeException("Bad or missing " + key + " in " + hint, e);
    }
}

// util

<T> void iterate(T state, Function<T,T> fn) {
    //noinspection InfiniteLoopStatement
    while(true) state = fn.apply(state);
}

void run(Path cwd, String... cmd) {
    if(unchecked(() -> new ProcessBuilder(cmd).directory(cwd.toFile()).inheritIO().start().waitFor()) != 0)
        throw new RuntimeException("Command failed: " + Arrays.toString(cmd));
}

// ----- app logic -----

void gitCloneRepo(Path root, String repo, String branch, String user) {
    run(Paths.get("."),"git","clone","--depth","1","--branch",branch,repo,root.toString());
    run(root, "git","config","user.email",user);
    run(root, "git","config","user.name", user);
}

void gitPull(Path root) {
    run(root, "git","pull");
}

void gitPush(Path root) {
    run(root, "git","add",".");
    run(root, "git","commit","-m","Sync updates");
    run(root, "git","push");
}

ArrayNode getUpdates(ObjectMapper mapper, String token, long offset) {
    final var uri = URI.create("https://api.telegram.org/bot" + token + "/getUpdates" + "?timeout=25&offset=" + offset);
    try (final var client = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(5)).build()) {
        final var req = HttpRequest.newBuilder(uri).timeout(Duration.ofSeconds(30)).GET().build();
        final var resp = unchecked(() -> client.send(req, HttpResponse.BodyHandlers.ofString()));
        if (resp.statusCode() != 200) throw new RuntimeException("getUpdates failed with HTTP " + resp.statusCode());
        final var payload = mapper.readTree(resp.body());
        if (!payload.get("ok").asBoolean()) throw new RuntimeException("Telegram getUpdates returned failure");
        return (ArrayNode) payload.get("result");
    }
}

record Msg(long chatId, String fileName, String body){}

void main(String[] args) {
    final var mapper = new ObjectMapper();
    final var conf = expect(k -> mapper.readTree(System.getenv(k)),"TG2SITE_CONF_CONTENT","env");
    final var secrets = expect(
            k -> mapper.readTree(unchecked(() -> Files.readString(Path.of(System.getenv(k))))),
            "TG2SITE_SECRETS_PATH","env"
    );
    final var repoURL   = expect(k -> secrets.get(k).asString(), "repository_url", "secret");
    final var branch    = expect(k -> conf.get(k).asString(),   "publish_branch", "conf");
    final var channelID = expect(k -> conf.get(k).asLong(),   "channel_id", "conf");
    final var token     = expect(k -> secrets.get(k).asString(),"telegram_token","secret");

    final var gitRoot = unchecked(() -> Files.createTempDirectory("tg2site-")).resolve("repo");
    System.err.println("cloning "+repoURL+" branch "+branch);
    gitCloneRepo(gitRoot, repoURL, branch, "bot@tg2site");
    final var msgDir = unchecked(() -> Files.createDirectories(gitRoot.resolve(".tg2site-messages")));
    final var msgKeys = List.of("channel_post", "edited_channel_post", "message", "edited_message");
    final var numR = Pattern.compile(".*/(\\d+).*\\.json$");
    final var msgFiles = using(() -> unchecked(()->Files.list(msgDir)), Stream::toList);
    final var maxId = msgFiles.stream().flatMap(path -> numR.matcher(path.toString()).results())
            .mapToLong(m -> Long.parseLong(m.group(1))).max().orElse(-1L);
    iterate(maxId, lastOffset -> {
        final var updates = getUpdates(mapper, token, lastOffset + 1);
        final var msgByChat = updates.valueStream().flatMap(upd -> msgKeys.stream().flatMap(mKey -> {
            final var message = upd.path(mKey);
            final var chatIdNode = message.path("chat").path("id");
            if (chatIdNode.isMissingNode()) return Stream.empty();
            final var path = "%020d.%s.json".formatted(upd.path("update_id").asLong(), mKey);
            final var body = mapper.writeValueAsString(message);
            return Stream.of(new Msg(chatIdNode.asLong(), path, body));
        })).collect(Collectors.groupingBy(Msg::chatId, Collectors.toList()));
        System.err.println("from chats: " + msgByChat.keySet().stream().sorted().toList());
        Stream.ofNullable(msgByChat.get(channelID)).forEach(relevant -> {
            gitPull(gitRoot);
            for (final var e: relevant)
                unchecked(() -> Files.writeString(gitRoot.resolve(".tg2site-messages").resolve(e.fileName), e.body));
            gitPush(gitRoot);
        });
        return updates.isEmpty() ? lastOffset : updates.get(updates.size()-1).get("update_id").asLong();
    });
}
